File: medium/space_kings.cpp
#include <string>
#include <iostream>
#include <sstream> 
#include <bits/stdc++.h>
// #include <vector>

using namespace std;
int max_gold=0;

void hit(vector<vector<int> > arr, int pointsSoFar, bool isAsha,int p,int q)
{
    if (arr.size() < 1)
    {
        max_gold = max(max_gold, pointsSoFar);
        return;
    }

    vector<vector<int> > a = arr;
    a[0][0] -= q;
    if (a[0][0] < 1)
    {
        a.erase(a.begin());
    }
    hit(a, pointsSoFar, true,p,q);

    if (isAsha)
    {
        for (int i = 0; i < arr.size(); i++)
        {
            vector<vector<int> > a = arr;
            a[i][0] -= p;
            int t = pointsSoFar;
            if (a[i][0] < 1)
            {
                t += a[i][1];
                a.erase(a.begin()+i);
            }
            hit(a, t, false,p,q);
        }
    }
}

int main()
{
    int p,q,n;
    cin>>p>>q>>n;
    max_gold=0;
    vector<vector<int> > arr( n , vector<int> (2, 0));
    
    for(int i=0;i<n;i++){
        int hp,score;
        cin>>hp>>score;
        arr[i][0]=hp;
        arr[i][1]=score;
    };
    // for (int i=0;i<n;i++){
    //     cout<<arr[i][0]<<" "<<arr[i][1]<<"\n";
    // };
    hit(arr,0,true,p,q);
    cout<<max_gold;
    return 0;
}

File: medium/space_kings.py
import copy
p=20
q=60
n=3

# alien_lives=[80,80,120]
# golds=[100,200,300]
arr=[
    [80,100],
    [80,200],
    [120,300]
]
# p=50
# q=400
# n=2
# arr=[
#     [60,100],
#     [190,200]
# ]

p=200
q=10
n=5
arr=[
    [20,100],
    [30,20],
    [20,101],
    [200,500],
    [10,1000]
]

max_gold=0

def hit(arr,pointsSoFar,isAsha):
    global max_gold
    # print(arr,pointsSoFar)
    if len(arr)==0:
        max_gold=max(max_gold,pointsSoFar)
        return
    
    # Asha skips, so Amir hits [or] just Amir hits
    a= copy.deepcopy(arr)
    a[0][0]-=q
    if a[0][0]<1:
        a.pop(0)
    hit(a,pointsSoFar,True)
    
    if isAsha:
        for i in range(len(arr)):
            a=copy.deepcopy(arr)
            a[i][0]-=p
            tempPoints=pointsSoFar
            if a[i][0]<1:
                tempPoints+=a[i][1]
                a.pop(i)
            hit(a,tempPoints,False)

hit(arr,0,True)
print(max_gold)